File: mostwater.java
public class mostwater{
    public static void main(String[] args) {
        int[] height = {1,8,6,2,5,4,8,3,7};
        System.out.println(maxArea(height));
    }

    public static int maxArea(int[] height){
        int left = 0;
        int right = height.length-1;
        int area = 0;
        int currentArea = 0;
        while(left < right){
            currentArea = Math.min(height[left], height[right]) * (right - left);
            if(area < currentArea)
                area = currentArea;
            if(height[left] > height[right])
                right--;
            else
                left++;
        }

        return area;
     }
}

//O(n) solution, only iterate through the list once.

File: mostwater.py
height = [1,8,6,2,5,4,8,3,7]
def most_water(lst):
    #pointer
    left = 0
    right = len(lst)-1
    curr_area = 0
    area = 0
    while left < right:
        #area formula
        curr_area = min(lst[left],lst[right]) * (right - left)
        area = max(area,curr_area)
        if lst[left] > lst[right]:
            right-=1
        else:
            left+=1
    return area

print(most_water(height))
#O(n) solution: no need to loop, just iterate through the list of heights. 